C/services/common/include/python_plugin_handle.h

```diff
@@ -17,19 +17,91 @@
 #include <plugin_handle.h>
 #include <Python.h>
 
+typedef void* (*pluginResolveSymbolFn)(const char *);
+typedef void (*pluginCleanupFn)();
+
 /**
- * The PythonPluginHandle class is used to represent an interface to 
+ * The PythonPluginHandle class is the base class used to represent an interface to
  * a plugin that is available as a python script
  */
 class PythonPluginHandle : public PluginHandle
 {
 	public:
-		PythonPluginHandle(const char *name, const char *path);
-		~PythonPluginHandle();
-		void *GetInfo();
-		void *ResolveSymbol(const char* sym);
+		// Base constructor
+		PythonPluginHandle(const char *name, const char *path) : m_name(name) {};
+		/**
+		 * Base destructor
+		 *    - Call cleanup on python plugin interface
+		 *    - Close python plugin interface library handle
+		 */
+		~PythonPluginHandle()
+		{
+			pluginCleanupFn cleanupFn =
+				(pluginCleanupFn) dlsym(m_hndl, "PluginInterfaceCleanup");
+			if (cleanupFn == NULL)
+			{
+				// Unable to find PluginInterfaceCleanup entry point
+				Logger::getLogger()->error("Plugin library %s does not support %s function : %s",
+							   m_interfaceObjName.c_str(),
+							   "PluginInterfaceCleanup",
+							   dlerror());
+			}
+			else
+			{
+				cleanupFn();
+			}
+			dlclose(m_hndl);
+		};
+
+		/**
+		 * Gets function pointer from loaded python interface library that can
+		 * be invoked to call 'sym' function in python plugin
+		 *
+		 * @param    sym	The symbol to fetch
+		 */
+		void *ResolveSymbol(const char* sym)
+		{
+			pluginResolveSymbolFn resolvSymFn =
+				(pluginResolveSymbolFn) dlsym(m_hndl, "PluginInterfaceResolveSymbol");
+			if (resolvSymFn == NULL)
+			{
+				// Unable to find PluginInterfaceResolveSymbol entry point
+				Logger::getLogger()->error("Plugin library %s does not support "
+							   "%s function : %s",
+							   m_interfaceObjName,
+							   "PluginInterfaceResolveSymbol",
+							   dlerror());
+			}
```

Comment:
if "resolvSymFn" is NULL here, we cannot proceed with using it below.

Reply:
Adding a check ...